src/nnet3bin/nnet3-xvector-get-egs.cc

```diff
@@ -0,0 +1,238 @@
+        || !ConvertStringToInteger(label_str, &(chunk_info->label)))
+        KALDI_ERR << "Expected integer for output archive in range file.";
+
+      chunk_info->name = utt + "-" + start_frame_str + "-" + num_frames_str
+        + "-" + label_str;
+      unordered_map<std::string, std::vector<ChunkInfo*> >::iterator
+        got = utt_to_chunks->find(utt);
+
+      if (got == utt_to_chunks->end()) {
+        std::vector<ChunkInfo* > chunk_infos;
+        chunk_infos.push_back(chunk_info);
+        utt_to_chunks->insert(std::pair<std::string,
+          std::vector<ChunkInfo* > > (utt, chunk_infos));
+      } else {
+        got->second.push_back(chunk_info);
+      }
+    }
+  }
+}
+
+static void WriteExamples(const MatrixBase<BaseFloat> &feats,
+    const std::vector<ChunkInfo *> &chunks, const std::string &utt,
+    bool compress, int32 num_pdfs, int32 *num_egs_written,
+    std::vector<NnetExampleWriter *> *example_writers) {
+  for (std::vector<ChunkInfo *>::const_iterator it = chunks.begin();
+      it != chunks.end(); ++it) {
+    ChunkInfo *chunk = *it;
+    NnetExample eg;
+    int32 num_rows = feats.NumRows(),
+          feat_dim = feats.NumCols();
+    if (num_rows < chunk->num_frames) {
+      KALDI_WARN << "Unable to create examples for utterance " << utt
+                 << ". Requested chunk size of "
+                 << chunk->num_frames
+                 << " but utterance has only " << num_rows << " frames.";
+    } else {
+      // The requested chunk positions are approximate. It's possible
+      // that they slightly exceed the number of frames in the utterance.
+      // If that occurs, we can shift the chunks location back slightly.
+      int32 shift = std::min(0, num_rows - chunk->start_frame
+                                 - chunk->num_frames);
+      SubMatrix<BaseFloat> chunk_mat(feats, chunk->start_frame + shift,
+                                  chunk->num_frames, 0, feat_dim);
+      NnetIo nnet_input = NnetIo("input", 0, chunk_mat);
+      for (std::vector<Index>::iterator indx_it = nnet_input.indexes.begin();
+          indx_it != nnet_input.indexes.end(); ++indx_it)
+        indx_it->n = 0;
+
+      Posterior label;
+      std::vector<std::pair<int32, BaseFloat> > post;
+      post.push_back(std::pair<int32, BaseFloat>(chunk->label, 1.0));
+      label.push_back(post);
+      NnetExample eg;
+      eg.io.push_back(nnet_input);
+      eg.io.push_back(NnetIo("output", num_pdfs, 0, label));
+      if (compress)
+        eg.Compress();
+
+      if (chunk->output_archive_id >= example_writers->size())
+        KALDI_ERR << "Requested output index exceeds number of specified "
+                  << "output files.";
+      (*example_writers)[chunk->output_archive_id]->Write(
+                         chunk->name, eg);
+      (*num_egs_written) += 1;
+    }
+  }
+}
+
+// Delete the dynamically allocated memory.
+static void Cleanup(unordered_map<std::string,
```

Comment:
Instead of having its own function I'd prefer to just put this in its own code block ({...}) with
a comment like
  ` {  // Free memory`
It feels like it should be part of the main function.
For the vector you can use DeletePointers() rather than writing a loop.

Reply:
Actually you could modify the DeletePointers() template so that it works with any container, not just vectors.  You could template it on the STL container, e.g.
`template <typename C> `...
and internally do:
```
  // T is the pointer type that the container contains, e.g. T == int*.
  typedef typename C::value_type T;
``